### ISSUE TITLE
Docutils: `@.*` gets converted to `&` in static filenames

### ISSUE THREAD

**Issue by gergelypolonkai** [getpelican/pelican#2646]
I’m currently migrating my blog from Jekyll to Pelican and try hard to preserve old filenames. One of these is my PGP key called `gergely@polonkai.eu.asc`. In my about page i added

```
You can download it `here <{static}../gergely@polonkai.eu.asc>`_

.. include:: ../gergely@polonkai.eu.asc
   :code: text
```

The file gets included (the content appears on the page), the link on the generated page references the correct filename, but the file does not exist, and i get this warning:

```
WARNING: Cannot get modification stamp for /home/polesz/Projects/blog/content/gergely&
  | 	FileNotFoundError: [Errno 2] No such file or directory: '/home/polesz/Projects/blog/content/gergely&'
```

If i escape the `@` character as `%40` or using a backslash (`\@`) the problem remains.

**Comment by gergelypolonkai** [comment]
The problem happens in `Content._link_replacer`, where `m.group('value')` becomes `../gergely&#64;polonkai.eu.asc` which gets URL parsed to

```
ParseResult(scheme='', netloc='', path='../gergely&', params='', query='', fragment='64;polonkai.eu.asc')
```

As a workaround, if i add `gergely@polonkai.eu.asc` to `STATIC_PATHS`, it gets copied to the output directory (thus, the generated link works), but the warning persists.

**Comment by oulenz** [comment]
I traced this back to docutils. I.e., the `@` is replaced by `&#64;` when the file is read in.

What I don't know is whether `&#64;` is generally the best way to render `@` in html. If yes, we could probably handle `@` in filenames similarly to how we already handle spaces. If no, then there might be a way to force docutils to leave `@` intact.

**Comment by avaris** [comment]
Well, that's a bit odd behavior of docutils. `&#64` would be the **HTML** encoding of `@`, but in a link you should **URL** encode if you encode it at all. That would be `%40` instead.

That aside, I don't know why [this](https://github.com/getpelican/pelican/blob/master/pelican/contents.py#L278) does naive replacement instead of `urllib.parse.unquote`, which could handle `%40` as well. However, it won't solve this issue since docutils is doing a different encoding. We *could* also do `html.unescape`, but that just feels wrong.

**Comment by oulenz** [comment]
Do you know whether docutuls does URL encoding generally? I've tried to look for it but couldn't find it.

**Comment by gergelypolonkai** [comment]
@avaris in that case, should i report this as a docutils bug?

**Comment by gergelypolonkai** [comment]
Can/should this be fixed on pelicans side, or should it be reported upstream in docutils instead?

**Comment by justinmayer** [comment]
@gergelypolonkai: I suggest reporting this to Docutils.

For whatever value it may have, I found several references to Docutils encoding the `@` symbol to `&#64;`:

https://chromium.googlesource.com/infra/third_party/docutils/+/refs/heads/master/docutils/docutils/writers/_html_base.py#290

https://docutils.sourceforge.io/docs/user/config.html#id137

From the changelog, back in 2003:
> Added `@` to `&#64;` encoding to thwart address harvesters.

**Comment by gergelypolonkai** [comment]
It seems that this mangling of email addresses as done only if some `cloak_email_addresses` setting is enabled (however it can be done) *and* it’s in a `mailto:` link:

```
    def visit_Text(self, node):
        text = node.astext()
        encoded = self.encode(text)
        if self.in_mailto and self.settings.cloak_email_addresses:
            encoded = self.cloak_email(encoded)
        self.body.append(encoded)
```

Off to the docutils issue board, then.

**Comment by avaris** [comment]
Based on the [response](https://sourceforge.net/p/docutils/bugs/387/), seems like *it's a feature, not a bug* ™. We can extend the search alternatives to HTML unescaped versions.

**Comment by justinmayer** [comment]
Thanks to @avaris working on the above-linked PR, this issue should now be addressed. Feel free to test latest master and post a comment here if any follow-up changes are deemed to be warranted.

**Comment by justinmayer** [comment]
Fix for this issue is included in the just-released [Pelican 4.5.1](https://github.com/getpelican/pelican/releases/tag/4.5.1). ✨
